zeebe/broker-client/src/main/java/io/camunda/zeebe/broker/client/impl/RoundRobinDispatchStrategy.java

```diff
@@ -10,30 +10,98 @@
 import io.camunda.zeebe.broker.client.api.BrokerClusterState;
 import io.camunda.zeebe.broker.client.api.BrokerTopologyManager;
 import io.camunda.zeebe.broker.client.api.RequestDispatchStrategy;
+import io.camunda.zeebe.dynamic.config.state.RoutingState;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * Return the next partition using a round-robin strategy, but skips the partitions where there is
  * no leader at the moment.
  */
 public final class RoundRobinDispatchStrategy implements RequestDispatchStrategy {
 
-  private final AtomicInteger partitions = new AtomicInteger(0);
+  /** Holds the current partition ring. Starts off uninitialized and is updated on every request. */
+  private final AtomicReference<VersionedPartitionRing> partitionRing =
+      new AtomicReference<>(VersionedPartitionRing.uninitialized());
+
+  private final AtomicInteger offset = new AtomicInteger(0);
 
   @Override
   public int determinePartition(final BrokerTopologyManager topologyManager) {
     final BrokerClusterState topology = topologyManager.getTopology();
 
-    if (topology != null) {
-      for (int i = 0; i < topology.getPartitionsCount(); i++) {
-        final int offset = partitions.getAndIncrement();
-        final int partition = topology.getPartition(offset);
-        if (topology.getLeaderForPartition(partition) != BrokerClusterState.NODE_ID_NULL) {
-          return partition;
-        }
+    if (topology == null) {
+      return BrokerClusterState.PARTITION_ID_NULL;
+    }
+
+    final var partitions = updatePartitionRing(topologyManager);
+
+    for (int i = 0; i < topology.getPartitionsCount(); i++) {
```

Comment:
Not really, the purpose of this loop is to try the next N partitions until we find a leader. It should _not_ iterate over `partitions.partitions()` because that would nullify the `offset` and prefer the first partitions heavily. Using `topology.getPartitionsCount()` as an upper bound looks a bit weird and we _could_ replace it with `partitions.partitions().size()` but it doesn't really matter.